### PR TITLE
Allow disabling the public-server-list

### PR DESCRIPTION
**Pull request by TimMThomas**
This implements #4070 and makes it possible to hide the public-server-list from the connect dialog.
The save location was left under ui for backwards compatibility.

This is a follow-up on #4305

### CODE REVIEW

**Comment by Popkornium18** [comment]
Not entirely sure but I think your changes are conflicting with mine which introduced some usages of the bool you renamed. I think you need to rebase from master and rename these too.

**Comment by TimMThomas** [comment]
Yea, saw that, im on it

**Comment by TimMThomas** [comment]
I think this should be it. However when i check the option and reopen the connection menu, mumble crashes. The Warning that the public list is disabled is printed. Can anyone replicate this?

**Comment by Popkornium18** [comment]
My guess is that since `qlPublicServers` is static it has still items once you reopen the connect dialog.

https://github.com/mumble-voip/mumble/blob/master/src/mumble/ConnectDialog.cpp#L1420

Then this gets executed because there are public servers in the list. But since `siPublic` has been set to `NULL` this leading to a crash.

Disabling the public list should probably also clear `qlPublicServers`.

**Comment by TimMThomas** [comment]
Jep, its a segmentation fault.
https://github.com/mumble-voip/mumble/blob/b1e8910c606c9f4e88e0f9016b54367643762696/src/mumble/ConnectDialog.cpp#L972-L986

if i add an "&& !g.s.bDisablePublicList" to the if it works fine. Would that also be solved by clearing the qlPublicServers? (Besides it being a good idea anyways)

**Comment by Popkornium18** [comment]
The text change is probably calling this slot.

https://github.com/mumble-voip/mumble/blob/b1e8910c606c9f4e88e0f9016b54367643762696/src/mumble/ConnectDialog.cpp#L1822-L1833

Which is then trying to filter the public server list and dereferencing the nullpointer here.

https://github.com/mumble-voip/mumble/blob/b1e8910c606c9f4e88e0f9016b54367643762696/src/mumble/ConnectDialog.cpp#L1523-L1525

Also it looks like I used spaces for tabs there :facepalm:

**Comment by TimMThomas** [comment]
So, i dont really know what the best approach would be. I add a toggled function to the checkbox, or i think it could work if i clear it in the NetworkConfig::save(). I dont really feel comfortable with both, any suggestions?

**Comment by Popkornium18** [comment]
I would check if the public list is disabled before calling `filterPublicServerList()` which solves the crashing issue. You could clear `qlPublicServers` when the box is checked. It's not strictly necessary since nothing is done with it if the public server list is disabled, but why keep it if it's not needed.

**Comment by TimMThomas** [comment]
alright, added the check, since im new to qt im still struggling with adding the action to the checkbox, I will try to do this on the weekend :)

**Comment by TimMThomas** [comment]
After messing a bit around i came to the conclusion that qlPublicServers should be anyway destroyed after the ConnectDialog closes/gets destroyed. So i think it shouldnt be necessary to do that. Then the pr should be ready like it is?

File: src/mumble/ConnectDialog.cpp [hunk 1]
```diff
@@ -1828,6 +1828,7 @@ void ConnectDialog::on_qcbFilter_currentIndexChanged(int filterIndex) {
 	} else if (filter == tr("Show Populated")) {
 		g.s.ssFilter = Settings::ShowPopulated;
 	}
-
-	filterPublicServerList();
+        if (!g.s.bDisablePublicList) {
+            filterPublicServerList();
+        }
```
Comment (Krzmbrzl) [inline src/mumble/ConnectDialog.cpp:1833]: I think it'd make more sense to add this check to `filterPublicServerList` itself than adding it here at this particular position.

Aka: The first thing I'd do inside `filterPublicServerList` would be to check whether the public server list is disabled and if it is, return immediately as in that case there is nothing to be filtered.
In that case you don't have to conditionally call the filter-function only when the list is enabled.

This has the advantage that if someone adds a call to the filter-function at another place in the future they can't forget to check that the list is actually enabled as that's done in the function's body.

File: src/mumble/NetworkConfig.ui [hunk 2]
```diff
@@ -340,6 +340,19 @@ Prevents the client from sending potentially identifying information about the o
...
+      <item>
+       <widget class="QCheckBox" name="qcbDisablePublicList">
+        <property name="enabled">
+         <bool>true</bool>
+        </property>
+        <property name="toolTip">
+         <string>When toggled this hides the public server list from the connect dialog</string>
+        </property>
+        <property name="text">
+         <string>Hide public-server-list</string>
```
Comment (Krzmbrzl) [inline src/mumble/NetworkConfig.ui:352]:
```suggestion
         <string>Hide public server list</string>
```
Otherwise the spelling would be inconsistent with the tooltip

File: src/mumble/Settings.cpp [hunk 3]
```diff
@@ -759,7 +759,8 @@ void Settings::load(QSettings* settings_ptr) {
 	SAVELOAD(bTCPCompat, "net/tcponly");
 	SAVELOAD(bQoS, "net/qos");
 	SAVELOAD(bReconnect, "net/reconnect");
-	SAVELOAD(bAutoConnect, "net/autoconnect");
+    	SAVELOAD(bAutoConnect, "net/autoconnect");
+    	SAVELOAD(bDisablePublicList, "ui/disablepubliclist");
```
Comment (Krzmbrzl) [inline src/mumble/Settings.cpp:763]: Please move this setting to the other ui-settings (where it was before)

File: src/mumble/Settings.cpp [hunk 4]
```diff
@@ -1115,6 +1115,7 @@ void Settings::save() {
 	SAVELOAD(bQoS, "net/qos");
 	SAVELOAD(bReconnect, "net/reconnect");
 	SAVELOAD(bAutoConnect, "net/autoconnect");
+	SAVELOAD(bDisablePublicList, "ui/disablepubliclist");
```
Comment (Krzmbrzl) [inline src/mumble/Settings.cpp:1118]: (see above)

File: src/mumble/Settings.h [hunk 5]
```diff
@@ -345,6 +345,7 @@ struct Settings {
 	bool bReconnect;
 	bool bAutoConnect;
 	bool bQoS;
+	bool bDisablePublicList;
```
Comment (Krzmbrzl) [inline src/mumble/Settings.h:348]:
```suggestion
	/// Disables the "Public Internet" section in the connect dialog if set.
	bool bDisablePublicList;
```
Why did you remove the comment in the first place?

**Comment by Krzmbrzl** [comment]
Oh and for your commit message, I think this slightly altered version would be better (plus I fixed a few typos):
```
FEAT(settings): Allow disabling of the public server list

This commit makes it possible to hide the public server list from the connect dialog. The setting's group was
left to be ui for backwards compatibility (even though the actual setting is part of the network settings in the UI).

Implements #4070
```

**Comment by Popkornium18** [comment]
> After messing a bit around i came to the conclusion that qlPublicServers should be anyway destroyed after the ConnectDialog closes/gets destroyed. So i think it shouldnt be necessary to do that. Then the pr should be ready like it is? 

It shouldn't be cleared when the dialog gets closed as that would require the server list to be downloaded every time the connect dialog is opened, which is unneccessary. But as I said, it isn't a big deal. It's no privacy risk, it's just some memory that could be freed. But once the client is restarted there will be no difference as the list will not get populated in the first place.

**Comment by Krzmbrzl** [comment]
I guess the saved memory is not noticeable on any system though, so I think that's okay :)

File: src/mumble/Settings.h (hunk 5 above)
Comment (TimMThomas) [reply src/mumble/Settings.h:348]: Tbh, i dont even know anymore that i removed that...

File: src/mumble/ConnectDialog.cpp (hunk 1 above)
Comment (TimMThomas) [reply src/mumble/ConnectDialog.cpp:1833]: Agreed

**Comment by TimMThomas** [comment]
Added the changes you requested. And yep, i feel like im blind, didnt saw that the list is static :sweat_smile:
On another note: thank you guys for your help :)

**Comment by Krzmbrzl** [comment]
We had some issues with the macOS CI (see #4349) - I rebased your branch against the current master brach and now all CIs should succeed. After that I think we can merge this PR :)

**Comment by Krzmbrzl** [comment]
(I also fixed some whitespace issue while I was on it - you had mixed indentation with tabs and with spaces)

**Comment by Krzmbrzl** [comment]
Thank you for your contribution! :)

**Comment by TimMThomas** [comment]
Alright, thank you all for you help! :)

**Comment by Krzmbrzl** [comment]
You're welcome :)
